// @sudoku/stores/session.js
// Store 适配层:将领域对象桥接到 Svelte 响应式系统
// 闭包中持有 Game 实例,对外暴露 writable store 快照
// HW2:在快照中加入 isExploring / failedHere / nextHint;新增 enterExplore / commit / abandon API
import { decodeSencode } from '@sudoku/sencode';
import { generateSudoku, solveSudoku } from '@sudoku/sudoku';
import { writable } from 'svelte/store';
import { createGame, createSudoku } from '../../../domain/index.js';

function createEmptyGrid() {
	return Array.from({ length: 9 }, () => Array(9).fill(0));
}

// 从 Game 实例生成 plain object 快照,供 Svelte store 消费
// 候选数 map 不放快照(81 格 × 9 候选过大,改为同步方法按需查询)
function createSnapshot(game) {
	return {
		baseGrid:     game.getInitialGrid(),
		userGrid:     game.getGrid(),
		invalidCells: game.getInvalidCells().map(({ row, col }) => `${col},${row}`),
		canUndo:      game.canUndo(),
		canRedo:      game.canRedo(),
		won:          game.isSolved(),
		// HW2 新增字段
		isExploring:  game.isExploring(),
		failedHere:   game.hasVisitedFailedState(),
		nextHint:     game.getNextHint(), // {row,col,value,reason} 或 null
	};
}

function createGameSession() {
	const session = writable(null);
	let game = createGame({
		sudoku: createSudoku(createEmptyGrid()),
	});

	// 每次领域变更后调用,生成新快照推送到 store 触发 UI 更新
	function publish() {
		session.set(createSnapshot(game));
	}

	function loadPuzzle(puzzleGrid) {
		game = createGame({
			sudoku: createSudoku(puzzleGrid),
		});
		publish();
		return true;
	}

	function guess(move) {
		const changed = game.guess(move);

		if (changed) {
			publish();
		}

		return changed;
	}

	function undo() {
		const changed = game.undo();

		if (changed) {
			publish();
		}

		return changed;
	}

	function redo() {
		const changed = game.redo();

		if (changed) {
			publish();
		}

		return changed;
	}

	// HW2:Hint 系列 API
	// applyHint 策略:
	//   1) 若该格只有一个候选数,走领域层 game.applyHint(纯推理,不依赖求解器)
	//   2) 否则回退到求解器 solveSudoku 找出该格答案,再走 game.guess
	// 两条路径都进入 Game 的 history,UI 的 Undo/Redo 一致工作
	function applyHint(position) {
		// position: {x, y} (与原 UI 兼容) 或 {row, col}
		const row = position.row ?? position.y;
		const col = position.col ?? position.x;

		// 优先用领域层推定
		const move = game.applyHint({ row, col });
		if (move) {
			publish();
			return true;
		}

		// 回退:基于初始题面求解(题面在游戏生命周期内不变,确保 Hint 不被用户错误填值污染)
		// 评审修复 2:原实现 solveSudoku(game.getGrid()) 用当前局面,用户填错后 solver 可能找不到解或给错值
		try {
			const solved = solveSudoku(game.getInitialGrid());
			const ok = game.guess({ row, col, value: solved[row][col] });
			if (ok) publish();
			return ok;
		} catch {
			return false;
		}
	}

	// HW2:Explore 系列 API
	function enterExplore() {
		const ok = game.enterExplore();
		if (ok) publish();
		return ok;
	}

	function commitExplore() {
		const ok = game.exitExplore('commit');
		if (ok) publish();
		return ok;
	}

	function abandonExplore() {
		const ok = game.exitExplore('abandon');
		if (ok) publish();
		return ok;
	}

	function resetExplore() {
		const ok = game.resetExplore();
		if (ok) publish();
		return ok;
	}

	publish();

	return {
		subscribe: session.subscribe,

		loadPuzzle,

		generate(difficulty) {
			return loadPuzzle(generateSudoku(difficulty));
		},

		decodeSencode(sencode) {
			return loadPuzzle(decodeSencode(sencode));
		},

		guess,

		applyHint,

		undo,

		redo,

		// HW2:候选数同步查询(不进快照,UI 按需调用)
		getCandidates(row, col) {
			return game.getCandidates(row, col);
		},

		getAllCandidates() {
			return game.getAllCandidates();
		},

		getNextHint() {
			return game.getNextHint();
		},

		// HW2:Explore 控制
		enterExplore,
		commitExplore,
		abandonExplore,
		resetExplore,
	};
}

export const gameSession = createGameSession();
